db/db_impl.cc

```diff
@@ -277,39 +311,57 @@ Status DBImpl::Resume() {
   }
   job_context.Clean();
 
-  ROCKS_LOG_INFO(immutable_db_options_.info_log, "Successfully resumed DB");
+  if (s.ok()) {
+    ROCKS_LOG_INFO(immutable_db_options_.info_log, "Successfully resumed DB");
+  }
   mutex_.Lock();
-  MaybeScheduleFlushOrCompaction();
+  // Check for shutdown again before scheduling further compactions,
+  // since we released and re-acquired the lock above
+  shutdown = shutting_down_.load(std::memory_order_acquire);
+  if (shutdown) {
+    s = Status::ShutdownInProgress();
+  }
+  if (s.ok()) {
+    for (auto cfd : *versions_->GetColumnFamilySet()) {
+      SchedulePendingCompaction(cfd);
+    }
+    MaybeScheduleFlushOrCompaction();
+  }
+
+  // Wake up any waiters - in this case, it could be the shutdown thread
+  bg_cv_.SignalAll();
 
   // No need to check BGError again. If something happened, event listener would be
   // notified and the operation causing it would have failed
-  return Status::OK();
+  return s;
 }
 
 // Will lock the mutex_,  will wait for completion if wait is true
-void DBImpl::CancelAllBackgroundWork(bool wait) {
+void DBImpl::CancelAllBackgroundWork(bool wait, bool shutdown) {
   InstrumentedMutexLock l(&mutex_);
 
-  ROCKS_LOG_INFO(immutable_db_options_.info_log,
-                 "Shutdown: canceling all background work");
+  if (shutdown) {
+    ROCKS_LOG_INFO(immutable_db_options_.info_log,
+                   "Shutdown: canceling all background work");
```

Comment:
If we set a flag here, will the race condition between recovery and shutdown be easier to resolve?

Reply:
Do you mean move up setting the ```shutting_down_``` flag here, or introduce a new flag?